mam0lyd Logout from account

mansurissa opened the pull request:
### WHAT DOES THIS PR DO?
This PR adds **Logout** functionality to phantom
---
### DESCRIPTION OF TASK TO BE COMPLETED
The user can login and be able to access his respective route(permission accordingly)
User can logout from account

### HOW SHOULD THIS BE MANUALLY TESTED?
- Do `npm install` 
- Create `.env` file and put configs as in `.env.example`
-Create three databases: `phantom-dev` , `phantom-test`, `phantom` 
-Do `npm start ` 
**Testing using postman**

- Make a POST request to the bus `/api/users/signin` use email from seeds and password
- Make a POST request to the bus `/api/users/{id}/logout` add the id of user and token to logout

### What are relevant trello story
mam0lyd

ghost commented:
<img src="https://www.deepcode.ai/icons/green_check.svg" width= "50px" align= "left"/> Congratulations :tada:. DeepCode [analyzed](https://www.deepcode.ai/app/gh/atlp-rwanda/incredible-phantom-backend/451e9d3229946911d3309f6f012b5abafc5dc944/atlp-rwanda/incredible-phantom-backend/bce54491cd59187d1487e2da66fe77627ffd4730/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) your code in 12.887 seconds and we found no issues. Enjoy a moment of no bugs :sunny:.

#### 👉 View analysis in [**DeepCode’s Dashboard**](https://www.deepcode.ai/app/gh/atlp-rwanda/incredible-phantom-backend/451e9d3229946911d3309f6f012b5abafc5dc944/atlp-rwanda/incredible-phantom-backend/bce54491cd59187d1487e2da66fe77627ffd4730/pr/_/%2F/code/?utm_source=gh_review&c=0&w=0&i=0&) | [_Configure the bot_](https://www.deepcode.ai/app/gh/?ownerconfig=atlp-rwanda)

mansurissa committed:
ft(logout): logout from account

- install and use redis
- set token in db and deleting
- testing
- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user
- adding translations to responses
- adding email template
- correcting heroku url

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Using google cloud transalte api for supporting multi languages

Changed to ES6 syntax

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix-issue-package-lock

feat(Roles & Permissions): Setting Roles and Permissions

-Admin can create,update,delete roles
-Admin can set and update Roles's permission

[Delivers #P4Z442wG]
- Ntare Guy -

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Using google cloud transalte api for supporting multi languages

Changed to ES6 syntax

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

feat(register) registring users

- random password
- sending email(not yet done)

 generating random password

 sending email

verify endpoint

 creating midleware for checking token

feat:(register): registering users

- registering user
- sending email after creating user
- sending comfirmation email after verifying account

ElYrZR34

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

ft-setup(register): register and roles$ permissions

- registering user
- hashing password
- protecting routes

feat(authentication):Register roles and permissions

- create user model and migration
- create seeder users
- protecting routes
- sign in
- asign token to signed in user
- making validators middleware

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Using google cloud transalte api for supporting multi languages

Changed to ES6 syntax

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix-issue-package-lock

feat(Roles & Permissions): Setting Roles and Permissions

-Admin can create,update,delete roles
-Admin can set and update Roles's permission

[Delivers #P4Z442wG]
- Ntare Guy -

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Using google cloud transalte api for supporting multi languages

Changed to ES6 syntax

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

feat(register) registring users

- random password
- sending email(not yet done)

 generating random password

 sending email

verify endpoint

 creating midleware for checking token

feat:(register): registering users

- registering user
- sending email after creating user
- sending comfirmation email after verifying account

ElYrZR34

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

resolving confilicts

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix-issue-package-lock

feat(Roles & Permissions): Setting Roles and Permissions

-Admin can create,update,delete roles
-Admin can set and update Roles's permission

[Delivers #P4Z442wG]
- Ntare Guy -

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

Fix-issue-package-lock

feat(Roles & Permissions): Setting Roles and Permissions

-Admin can create,update,delete roles
-Admin can set and update Roles's permission

[Delivers #P4Z442wG]
- Ntare Guy -

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

Using google cloud transalte api for supporting multi languages

Changed to ES6 syntax

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-CRUD-role

CREATE Roles
READ Roles
UPDATE Roles
DELETE Roles

ch-permissions-middleware

Fix issue #15

Fixed Issue_#15'

issue_#15

Fixed issue #15 and i18n setup

xUjJi1sq Chore Support multiple languages

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

ch-setup(initial setup): setting up project

- install express
- install dotenv
- making test route

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

feat(register) registring users

- random password
- sending email(not yet done)

 generating random password

 sending email

verify endpoint

 creating midleware for checking token

feat:(register): registering users

- registering user
- sending email after creating user
- sending comfirmation email after verifying account

ElYrZR34

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

feat(authorization):Register and routes protection

- register user
- get all users
- create role
- protecting routes
- permissions according to the role

 adding travis config for database and tests

 adding documentation

feat(authentication):signin

- create user model and migration
- create seeder users
- hash password
- sign in
- asign token to signed in user

chore:(eslint): configuringneslint

- install and configure eslint
- using airbnb base style

chore: connect-to-db
- creating seeds
- creating migration
- user model

 bg-fixing some model error

ft: signin

- registering user
- validations with join

 hashing pass2word before saving in db

feat(authenticatiom):sign in

- login for user
- signing token to signed in user

 ft(logout): logout from account
- user can logout
- token get deleted from redis server
- tests
- using redis to store sessions
- used remote redis

41hulk reviewed (commented):
Great Work adding reddis

raymond42 commented on .env.example:
```diff
@@ -1,9 +1,11 @@
-JWT_KEY=<jwt key>
-EMAIL=<email>
-PASS=<password>
+DEV_DATABASE_URL=postgres://postgres:pass2db@localhost:5432/phantom-dev
+TEST_DATABASE_URL=postgres://postgres:pass2db@localhost:5432/phantom-test
+DATABASE_URL=postgres://postgres:pass2db@localhost:5432/phantom
+JWT_KEY=helloPhantom
+EMAIL=atlpincredible@gmail.com
+PASS=andela@2020
```
you didn't have to change the values for DEV_DATABASE_URL, TEST_DATABASE_URL, DATABASE_URL, JWT_KEY, EMAIL

raymond42 commented on src/configs/redisClient.js:
```diff
@@ -0,0 +1,13 @@
+config();
+
+const redisClient = redis.createClient({
+  url: process.env.REDIS_URL,
+  password: process.env.REDIS_PWD,
+});
```
Destructure REDIS_URL, and REDIS_PWD from process.env

mansurissa replied:
Done, thanks

raymond42 commented on src/utils/mail2.js:
```diff
@@ -1,36 +1,73 @@
+      product: {
+        name: 'Phantom',
+        link: `${process.env.HOST}`,
+        // logo: 'https://mailgen.js/img/logo.png'
```
remove this commented line

mansurissa replied:
Done

raymond42 reviewed (changes requested):
Good work so far. Please attend to the comments left and revert to me

mansurissa commented:
> Good work so far. Please attend to the comments left and revert to me

Thanks.

raymond42 reviewed (approved):
LGTM
